presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -383,26 +388,57 @@ private synchronized Optional<QueryResults> getCachedResult(long token)
     private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uriInfo, String scheme, DataSize targetResultSize, boolean binaryResults)
     {
         QueryResults queryResults = getNextResult(token, uriInfo, scheme, targetResultSize, binaryResults);
-        if (queryResults.getError() == null || !queryResults.getError().isRetriable()) {
-            return queryResults;
-        }
 
-        // check if we have exceeded the global limit
-        retryCircuitBreaker.incrementFailure();
-        if (!retryCircuitBreaker.isRetryAllowed() || hasProducedResult) {
+        boolean historyBasedOptimizationEnabled = useHistoryBasedPlanStatisticsEnabled(session) && trackHistoryBasedPlanStatisticsEnabled(session);
+        boolean underOneRetryCount = queryManager.getQueryRetryCount(queryId) < 1;
+        boolean underRetryExecutionTime = queryManager.getQueryInfo(queryId).getQueryStats().getExecutionTime().toMillis() <= getQueryRetryMaxExecutionTime(session).toMillis();
+        boolean planHashesAreSame = true;
+
+        List<CanonicalPlanWithInfo> currentPlanHashes = queryManager.getFullQueryInfo(queryId).getPlanCanonicalInfo();
+
+        if (queryResults.getError() != null && historyBasedOptimizationEnabled && underOneRetryCount && underRetryExecutionTime && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            if (lastPlanHashes == null) {
+                lastPlanHashes = currentPlanHashes;
+            }
+        }
+        else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            if (currentPlanHashes.size() == lastPlanHashes.size()) {
+                for (int planHash = 0; planHash < currentPlanHashes.size(); planHash++) {
+                    if (!currentPlanHashes.get(planHash).getInfo().getHash().equals(lastPlanHashes.get(planHash).getInfo().getHash())) {
```

Comment:
Note Don't think its a deal breaker but the plan hashes can change if the underlying data has changed. (if someone added a new partition)

Reply:
Thinking of potentially future work that can be built of this. For the first version I wanted to keep it simple, but I agree with you. [Plan Hash vs Plan comparison doc](https://docs.google.com/document/d/1w6yQojt6wLfs509uXNaoAx1tNgWw8inIsNARsawLeQg/edit?usp=sharing)